doc/lbrc.1
.TH lbrc 1 "August 2005"  "Version 1.0.0" "1.0.0"
.SH NAME
lbrc \- Configuration file for execution of lb

.SH DESCRIPTION

The
.B lbrc
file controls the way that the
.B lb
program generates reports.  The lbrc file defines the sections and
arithmatic for the various accounts that it will be reporting on.
There is a default configuration file that is compiled into the
.B lb
program, in case you do not define an lbrc file yourself.

.SH SEARCH PATH

The
.B lb
and
.B lbgui
programs will search for the following files (in order) before falling
back on the default configuration:

.IP 1:
 ./.lbrc
.IP 2:
 ~/.lbrc

.P
If no configuration file is found, a default one is supplied.  See the
.B EXAMPLE
section.

.SH CONFIGURATION

The lbrc file is like a blueprint for your report.  The report will be
ouput in roughly the same order that you had in your configuration
file.  Blank lines and lines that begin with '#' are ignored.  There
are only two types of configuration command:  
.B section
and
.B report.

.SS section

A section is a reporting of each individual account.  It finds
accounts that match your criteria, and then groups them together.  The
line looks like this:
.P
section <"n"> <"Label"> <"Match"> [-]1

Each field is described thus:

.TP 12
.BI section
Denotes that we're defining a section
.TP 12
.BI ``n''
A
.I variable
name to be used later in a
.B report
field.  Use a single letter.  Case sensitive.
.TP 12
.BI ``Label''
A string that will announce the summary of these files (at the bottom
line).  E.g. if these are 'expense accounts' you would probably want
to use something like ``Total Expenses:''.
.TP 12
.BI ``Match''
Matches this string with the name of each account in the budgets and
reports.  If the first part of the actual account matches this string,
then the account will be reported in this section.  Note that accounts
should not appear in more than one section.
.B IMPORTANT!
You should always have the last 'section' be "", which makes it a
default handler.  Reports
.I will not
be reported twice.  This will print out all accounts that have not
been reported in previous sections.
.TP 12
.BI [-]1
This should either be '1' or '-1'.  If the account is an
income type of account, the value should be '1'.  If the account is an
expense type of account, the value should be '-1'.  This way, when the
report comes on an account, the values are generally spit out as
non-negative numbers (which is easier to understand).

.SS report

A report gives a sort of bottom line statement (like ``YE OLDE BOTTOM
LINE:'') and performs a little arithmatic.  The format of a 'report'
line is like this:
.P
report <"Label"> <"sums">

.TP 12
.BI report
Denotes that we're defining a report.
.TP 12
.BI ``Label''
An arbitrary introduction to the line.
.TP 12
.BI ``sums''
A very simple equation.  It only handles + and -, no parenthesis.
E.g. if you have defined sections a, b, and c previously, then you may
enter the equation as ``a-b-c'' or ``a+b-c'', etc.  You may not get
any more complicated than that.  E.g. ``a-(c-b)'' will give
unpredictible results.

.SH EXAMPLES

.SS The default settings

When no lbrc file is found, this is the file that is used:

.PD 0
.P
section "a" "Total Income:" "Income:" 1
.P
section "b" "Total Expenses:" "" -1
.P
report "Bottom line:" "a-b"

All accounts that begin with ``Income:'' (E.g. ``Income:Salary'' and
``Income:Other'') will be reported in this section.  Their values will
be summed at the bottom of the section with a label that says ``Total
Income:''.  Since it's income, we use the 1 to denote a positive
cash-flow.  The second section uses the default ``Match'' parameter,
so it will report 
.I every
account that was not reported as Income.  Presuming that they are all
expenses (a negative cash flow), we indicate with a -1 at the end of
the line.

The ``Bottom line'' report merely shows how much you have left over
(or how far you have fallen behind) when the expenses (``b'') are
subtracted from the income (``a'').

.SS My settings

Here's the lbrc file that I use:

.PD 0
.P
section "a" "Income Accounts:" "Income:" 1
.P
section "b" "Savings Accounts:" "Assets:Current Assets:Savings Account" -1
.P
section "c" "Debts:" "Liabilities:" -1
.P
section "d" "Expenses:" "" -1
.P
report "Bottom line (without debts):" "a-b-c-d"

This is about the same as above, but I have certain types of accounts
(savings and debts) laid out in their own sections.

.SH "SEE ALSO"

.BR lb (1),
.BR lbgui (1)

.SH AUTHOR

G. M. Beddingfield <gabriel@teuton.org>


ChangeLog
ChangeLog for lb
	
2005-10-28 Gabriel M. Beddingfield <gabriel@teuton.org>
	* Release:  4.0.1
	* Fixed build errors on gcc4 / FC4
	* Updated autoconf tools.


src/core/lbrc.cpp
/***************************************************************************
 *                                                                         *
 *   LB - The Little Budget Utility for GnuCash                            *
 *                                                                         *
 *   Copyright (C) 2004,2005 by Gabriel M. Beddingfield                    *
 *   gabriel@teuton.org                                                    *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

/* lbrc.cpp
   Gabriel M. Beddingfield
   2005-08-09

   This class manages the global settings for lb.
*/

#include <cassert>
#include <string>
#include <sstream>
#include <iostream>
#include <iomanip>
#include <ios>
#include <fstream>
#include <cstdlib> // for getenv()
#include "lbrc.h"
#include "lb.h" // for lb_error
#include "bud.h"

using namespace std;

template<class A, class B, class C, class D, class E, class F, class G>
inline static string ONELINE(A _a,B _b,C _c,D _d,E _e,F _f,G _g)
{
	ostringstream os;
	os << setw(37) << (_a) << "  "				     
	   << setw(9) << setprecision(2) 
	   << setiosflags(ios::right|ios::fixed) 
	   << (_b) << " "	
	   << setw(9) << (_c) << " " 
	   << setw(9) << (_d) << " | " 
	   << setw(9) << (_e) << " " 
	   << setw(9) << (_f) << " " 
	   << setw(9) << (_g) 
	   << endl;
	return os.str();
}




/** \brief Determines if the given file exists.
 *
 */

inline static bool exists( const string& fn )
{
	ifstream fs;
	fs.open( fn.c_str(), ios::in );
	return fs.is_open();
}

/** \brief Gives the name of the resource file.
 *
 * Returns the filename with the complete path of the files location.
 * Right now it only gives either ./.lbrc or $HOME/.lbrc.  If neither
 * exists, it returns an empty string.
 *
 */

static string findfile(void)
{
	string fn(".lbrc");
	char* str;
	if( exists(fn) ) return fn;

	str = getenv("HOME");
	if( str != 0 ) {
		fn = string(str) +
			string("/") + fn;
		if( exists(fn) ) return fn;
	}
	return string();
}

lbrcItem::lbrcItem(lbrc* parent) : bud( 0.0, 0.0, 0.0, 0.0 ), _parent(parent)
{

}

lbrcItem::lbrcItem( float act, float plan, float ytd_act, float ytd_plan, lbrc* parent ) :
	bud( act, plan, ytd_act, ytd_plan ), _parent(parent)
{

}

lbrcItem::~lbrcItem(void)
{

}

lbrc::lbrc( const string& filename )
{
	init(filename);
}

/** \brief Same as default constructor.
 *
 */
lbrc::lbrc( const char* filename )
{
	init( string(filename) );
}

lbrc::~lbrc()
{

}

void lbrc::init( const string& filename )
{
	if( filename.empty() ) {
		parse_config( findfile() );
	} else {
		assert( exists(filename) );
		parse_config( filename );
	}
	return;
}

void lbrc::reset( const string& filename )
{
	clear();
	assert( size() == 0 );
	init( filename );
}
string lbrc::header( accounts* a )
{
	ostringstream os;
	os << endl;
	os << ONELINE("Account","Planned","Actual","Diff","YTD Plan","YTD Act","YTD Diff");
	os << ONELINE("=====================================",
		      "=========","=========","=========","=========","=========","=========");
	return os.str();
}

string lbrc::footer( accounts* a )
{
	return string();
}

string lbrcItem::oneline( accounts::iterator i )
{
	ostringstream os;
	// Match i->first(), if matches, then insert to the stream.
	if( i->second.flag() && match(i) ) {
		os << ONELINE( i->first,
			       i->second.planned(),
			       _sign*i->second.actual(),
			       _sign*i->second.actual() - i->second.planned(),
			       i->second.ytd_planned(),
			       _sign*i->second.ytd_actual(),
			       _sign*i->second.ytd_actual() - i->second.ytd_planned() );
		i->second.clear();
		this->clear();
		// Add these sums to ours, for summary()
		planned( i->second.planned() );
		actual( _sign * i->second.actual() );
		ytd_planned( i->second.ytd_planned() - i->second.planned() );
		ytd_actual( _sign * (i->second.ytd_actual() - i->second.actual()) );
	}
	return os.str();
}

string lbrcItem::summary( accounts* a )
{
	// Two types:  Section and Report
	if( _type == lbrcItem::Section ) {
		return sectionSummary( a );
	} else {
		return reportSummary( a );
	}
}

string lbrcItem::sectionSummary( accounts* )
{
	if( flag() ) return string(); // i.e. no accounts were matched
	ostringstream os;
	os << ONELINE( "",
		       "=========",
		       "=========",
		       "=========",
		       "=========",
		       "=========",
		       "=========")
	   << ONELINE( _name,
		       planned(),
		       actual(),
		       diff(),
		       ytd_planned(),
		       ytd_actual(),
		       ytd_diff() )
	   << endl;
	return os.str();
}

string lbrcItem::reportSummary( accounts* )
{
	lbrc::iterator i;
	ostringstream os;
	size_t pos;
	unsigned char c;

	for( i=_parent->begin() ; &(*i)!=this ; ++i ) {
		pos = _match.find( i->_id );
		if( pos != string::npos ) {
			c = (pos==0)?'+':_match[pos-1];
			switch( c ) {
			case '+':
				planned( i->planned() );
				actual( i->actual() );
				ytd_planned( i->ytd_planned() - i->planned() );
				ytd_actual( i->ytd_actual() - i->actual() );
				break;
			case '-':
				planned( -i->planned() );
				actual( -i->actual() );
				ytd_planned( -i->ytd_planned() + i->planned() );
				ytd_actual( -i->ytd_actual() + i->actual() );
				break;
			}
		}
	}
	os << endl
	   << ONELINE( _name,
		       planned(),
		       actual(),
		       diff(),
		       ytd_planned(),
		       ytd_actual(),
		       ytd_diff() )
	   << endl;
	return os.str();
}


bool lbrcItem::match( accounts::iterator i )
{
	if( _match.length() == 0 ) {
		return i->second.flag();
	}
	return ( 0 == _match.compare( 0, _match.length(), i->first, 0, _match.length() ) ) ?
		 true : false;
}

void lbrc::parse_config( const string& filename )
{
	if( filename.empty() || !exists(filename) ) {
		default_config();
	} else {
		assert( exists(filename) );
		ifstream fp( filename.c_str(), ios::in );
		if( fp.bad() ) {
			default_config();
			return;
		}
		ostringstream os;
		os << fp.rdbuf();
		string s = os.str();
		if( s.empty() ) {
			default_config();
		} else {
			parse_config_text( s );
		}
	}

	return;
}

// Here's how I want to read in the configuration file...
// (open file)
// (while have lines) {
//       process the line
// }
//
// Line processing basically has 4 branches:
// + ignore (comment, whitespace)
// + warn -- same as ignore, but warn about bad syntax
// + Section (switch to warn on bad syntax)
// + Report (switch to warn on bad syntax)
//
// We'll still process the file as a whole from the string.
//

static size_t getOneField( const string& s,
			   size_t a,
			   string* x,
			   unsigned char* flag )
{
	size_t b;
	if( a == string::npos ) {
		*flag=2;
		return a;
	}
	a = s.find('\"',a);
	if( a == string::npos ) {
		*flag=2;
		return a;
	}
	++a;
	b = s.find('\"',a);
	if( b == string::npos ) {
		*flag=2;
		return b;
	}
	(*x) = s.substr(a,b-a);
	++b;
	return b;
}

static size_t getSign( const string& s,
		       size_t a,
		       char* x,
		       unsigned char* flag )
{
	size_t b;
	if( a == string::npos ) {
		*flag = 2;
		return a;
	}
	a = s.find_first_of("+-1");
	if( a == string::npos ) {
		*flag = 2;
		return a;
	}
	// Now, the string should be either "1", "+1", or "-1".
	b = s.find_first_not_of("+-1");
	string tmp = s.substr(a,b-a);
	if( tmp == "1" || tmp == "+1" ) {
		*x = 1;
	} else if ( tmp == "-1" ) {
		*x = -1;
	} else {
		*flag = 2;
	}
	return b+1;
}

unsigned char lbrcItem::line2lbrcItem( const string& ln,
				       string* err )
{
	unsigned char flag=0;
	const string ws = " \a\b\f\n\r\t\v";
	size_t a;

	// Ignore comments and whitespace.
	if( '#' == ln[0] ) return flag;
	if( ln.find_first_not_of(ws) == string::npos ) return flag;

	// Determine whether Section, Report, or other.
	if( 0 == ln.compare(0,6,"section",0,6) ) {
		// First part: name
		_type = Section;
		a = getOneField( ln, 7, &_id, &flag );
		a = getOneField( ln, a, &_name, &flag );
		a = getOneField( ln, a, &_match, &flag );
		a = getSign( ln, a, &_sign, &flag );
		if( 2 == flag ) {
			(*err) = string("Bad syntax ") + ln;
		} else {
			flag = 1;
		}
	} else if( 0 == ln.compare(0,5,"report",0,5) ) {
		// First part: name
		_type = Report;
		a = getOneField( ln, 6, &_name, &flag );
		a = getOneField( ln, a, &_match, &flag );
		if( 2 == flag ) {
			(*err) = string("Bad syntax ") + ln;
		} else {
			flag = 1;
		}
	} else {
		flag=2;
		(*err) = string("Bad syntax: ") + ln;
	}

	return flag;
}

void lbrc::parse_config_text( string& s )
{
	lbrcItem* x;
	size_t ptr;
	string ln;
	string err;
	unsigned char flag;
	int i=0;

	assert( !s.empty() );

	while( !s.empty() ) {
		x = new lbrcItem(this);
		++i;  // processing line 'i'
		ptr = s.find('\n');
		ln = s.substr(0,ptr);
		s.erase(0,ptr+1);
		flag = x->line2lbrcItem( ln, &err );
		switch( flag ) {
		case 0: // Ignored line.
			break;
		case 1: // Houston, we have data.
			push_back(*x);
			break;
		case 2: // Bad syntax - warn
			cerr << "config file line " << i 
			     << " " << err << endl;
			break;
		default:
			throw lb_error("Unhandled error in lbrc parse");
			break;
		}
		delete x;
	}
}

void lbrc::default_config( void )
{
	string s("section \"a\" \"Total Income:\" \"Income:\" 1\n"
		 "section \"b\" \"Total Expenses:\" \"\" -1\n"
		 "report \"Bottom line:\" \"a-b\"\n");

	parse_config_text( s );

	return;
}
